plugins/shared/loader/instance.go

```diff
@@ -0,0 +1,54 @@
+package loader
+
+import plugin "github.com/hashicorp/go-plugin"
+
+// PluginInstance wraps an instance of a plugin. If the plugin is external, it
+// provides methods to retrieve the ReattachConfig and to kill the plugin.
+type PluginInstance interface {
+	// Internal returns if the plugin is internal
+	Internal() bool
+
+	// Kill kills the plugin if it is external. It is safe to call on internal
+	// plugins.
+	Kill()
+
+	// ReattachConfig returns the ReattachConfig and whether the plugin is internal
+	// or not. If the second return value is true, no ReattachConfig is possible to
+	// return.
+	ReattachConfig() (*plugin.ReattachConfig, bool)
```

Comment:
The bool designating that it's internal isn't very intuitive. Maybe flipping the logic and making the bool something like `canReattach` where if it's false then no ReattachConfig is possible.